### DISCUSSION TITLE
`impl ToSql for std::net::IpAddr`?

### DISCUSSION THREAD

**Discussion by maltalef101** [diesel-rs/diesel#3434] [Q&A]
Rust already provides an enum for storing this type of data, be it IPv4 or IPv6 either `std::net::Ipv4Addr` or `std::net::Ipv6Addr` are used inside one of the two enum variants; so why isn't this used? I see that there are implementations for types from external crates for this type of data, and I also see that there implementations for other types in the standard library, so this case seems  weird to me.

Maybe there's a reason for this.

**Comment by weiznich** [comment]
The main reason why this is currently unsupported is that `std::net::IpAddr` does not support network masks, which are supported by the corresponding postgresql type.

See https://github.com/diesel-rs/diesel/pull/2647 and #749 for details

**Reply by Mingun** [reply]
From technical point of view implementing only `ToSql` (as asked) does not requiring supporting masks, because PG type can hold anything that `std::net::IpAddr` is able to hold. So the question is what the policy of diesel for not symmetric `ToSql` / `FromSql` implementations.

**Reply by weiznich** [reply]
We generally try to provide either both `ToSql` and `FromSql` or none of them. There are a few exceptions to this rule, but you would need to present a really good reason for getting another `ToSql`/`FromSql` implementation without matching counter part into diesel.

(Thats also outlined in one of the linked threads)

**Reply by maltalef101** [reply]
Ah, I see. There *is* a reason.

 @weiznich 's point of view seems reasonable to me. If you wanted to store that data, you should have some way of accessing it.

Thanks! 😀
